src/gpu/intel/jit/ir/core.cpp

```diff
@@ -64,6 +63,36 @@ int type_t::size() const {
     return 0;
 }
 
+int type_t::mantissa_bits() const {
+    if (!is_fp()) return 0;
+
+    switch (kind()) {
+        case type_kind_t::f64: return 52;
+        case type_kind_t::f32: return 23;
+        case type_kind_t::tf32:
+        case type_kind_t::f16: return 10;
+        case type_kind_t::bf16: return 7;
+        case type_kind_t::hf8: return 3;
+        case type_kind_t::bf8: return 2;
+        case type_kind_t::f4_e2m1: return 1;
+        case type_kind_t::f4_e3m0: return 0;
+        default: gpu_error_not_expected();
+    }
+    return 0;
+}
+
+bool is_subset(const type_t &a, const type_t &b) {
+    if (a.is_undef() || b.is_undef()) return false;
+    if (a == b) return true;
+    if (a.is_fp() && b.is_int()) return false;
+    if (a.is_x16() && b.is_tf32()) return false;
+    if (a.is_int() && b.is_fp())
+        return a.bitsize() <= b.mantissa_bits() + a.is_signed();
+    if (a.is_tf32() && b.is_f32()) return true;
+    if (a.is_fp() && b.is_fp() && a.bitsize() == b.bitsize()) return false;
+    return a.bitsize() <= b.bitsize();
```

Comment:
Nit: am I right that this will return `true` for `is_subset(u32, s32)`? Is it the intention? I'm asking because `s32` can't hold large `u32` values.